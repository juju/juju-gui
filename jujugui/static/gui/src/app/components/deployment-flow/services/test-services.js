/* Copyright (C) 2017 Canonical Ltd. */
'use strict';

const React = require('react');
const enzyme = require('enzyme');
const shapeup = require('shapeup');

const DeploymentServices = require('./services');

describe('DeploymentServices', function() {
  let acl, changesUtils;

  const renderComponent = (options = {}) => enzyme.shallow(
    <DeploymentServices
      acl={options.acl || acl}
      addNotification={options.addNotification || sinon.stub()}
      changesUtils={options.changesUtils || changesUtils}
      charmsGetById={options.charmsGetById || sinon.stub()}
      getCurrentChangeSet={options.getCurrentChangeSet || sinon.stub()}
      getServiceByName={options.getServiceByName || sinon.stub()}
      listPlansForCharm={options.listPlansForCharm || sinon.stub()}
      parseTermId={options.parseTermId || sinon.stub()}
      showTerms={options.showTerms || sinon.stub()}
      withPlans={options.withPlans} />
  );

  beforeEach(() => {
    acl = {isReadOnly: sinon.stub().returns(false)};
    changesUtils = shapeup.addReshape({
      generateAllChangeDescriptions: sinon.stub(),
      generateChangeDescription: sinon.stub(),
      sortDescriptionsByApplication: sinon.stub().returns(
        JSON.parse('{"kibana":[{"id":"service-131","icon":"https://api.jujucharms.com/charmstore/v5/trusty/kibana-15/icon.svg","description":" kibana will be added to the model.","time":"1:28 pm"},{"id":"addUnits-655","icon":"changes-units-added","description":" 1 kibana unit will be added.","time":"1:28 pm"}],"elasticsearch":[{"id":"setConfig-169","icon":"changes-config-changed","description":"Configuration values will be changed for elasticsearch.","time":"1:28 pm"}]}') // eslint-disable-line max-len
      )
    });
  });

  it('can render', function() {
    const wrapper = renderComponent({
      getCurrentChangeSet: sinon.stub().returns({}),
      withPlans: true
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('can render without plans', function() {
    const wrapper = renderComponent({
      getCurrentChangeSet: sinon.stub().returns({}),
      withPlans: false
    });
    assert.equal(wrapper.find('BudgetTable').prop('withPlans'), false);
    assert.equal(wrapper.find('.deployment-services__spend').length, 0);
  });

  it('can render with machines and no services', function() {
    const machineChanges = [
      {id: 'machine0'},
      {id: 'machine1'}
    ];
    changesUtils.generateChangeDescription.onFirstCall().returns(machineChanges[0]);
    changesUtils.generateChangeDescription.onSecondCall().returns(machineChanges[1]);
    changesUtils.sortDescriptionsByApplication.returns(null);
    const wrapper = renderComponent({
      getCurrentChangeSet: sinon.stub().returns({
        destroy1: {
          command: {
            method: '_destroyMachines'
          }
        },
        destroy2: {
          command: {
            method: '_addMachines'
          }
        }
      })
    });
    expect(wrapper).toMatchSnapshot();
  });
});
