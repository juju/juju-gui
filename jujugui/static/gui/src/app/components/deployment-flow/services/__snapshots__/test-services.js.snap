// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentServices can render 1`] = `
<div>
  <BudgetTable
    acl={
      Object {
        "isReadOnly": [Function],
      }
    }
    addNotification={[Function]}
    allocationEditable={true}
    charmsGetById={[Function]}
    extraInfo={
      Object {
        "elasticsearch": <ul
          className="deployment-services__changes"
        >
          <li>
            <DeploymentChangeItem
              change={
                Object {
                  "description": "Configuration values will be changed for elasticsearch.",
                  "icon": "changes-config-changed",
                  "id": "setConfig-169",
                  "time": "1:28 pm",
                }
              }
              showTime={false}
            />
          </li>
        </ul>,
        "kibana": <ul
          className="deployment-services__changes"
        >
          <li>
            <DeploymentChangeItem
              change={
                Object {
                  "description": " kibana will be added to the model.",
                  "icon": "https://api.jujucharms.com/charmstore/v5/trusty/kibana-15/icon.svg",
                  "id": "service-131",
                  "time": "1:28 pm",
                }
              }
              showTime={false}
            />
          </li>
          <li>
            <DeploymentChangeItem
              change={
                Object {
                  "description": " 1 kibana unit will be added.",
                  "icon": "changes-units-added",
                  "id": "addUnits-655",
                  "time": "1:28 pm",
                }
              }
              showTime={false}
            />
          </li>
        </ul>,
      }
    }
    listPlansForCharm={[Function]}
    parseTermId={[Function]}
    plansEditable={true}
    services={
      Array [
        undefined,
        undefined,
      ]
    }
    showTerms={[Function]}
    withPlans={true}
  />
  <div
    className="deployment-services__spend prepend-seven"
  >
    Maximum monthly spend:Â 
    <span
      className="deployment-services__max"
    >
      $100
    </span>
  </div>
</div>
`;

exports[`DeploymentServices can render with machines and no services 1`] = `
<div>
  <div
    className="v1"
  >
    <BasicTable
      headerClasses={Array []}
      headers={
        Array [
          Object {
            "content": "Machines",
          },
        ]
      }
      rowClasses={Array []}
      rows={
        Array [
          Object {
            "columns": Array [
              Object {
                "content": <DeploymentChangeItem
                  change={
                    Object {
                      "id": "machine0",
                    }
                  }
                  showTime={false}
                />,
              },
            ],
            "key": "machine0",
          },
          Object {
            "columns": Array [
              Object {
                "content": <DeploymentChangeItem
                  change={
                    Object {
                      "id": "machine1",
                    }
                  }
                  showTime={false}
                />,
              },
            ],
            "key": "machine1",
          },
        ]
      }
    />
  </div>
</div>
`;
